apps/passport-server/test/zuconnect.spec.ts

```diff
@@ -0,0 +1,242 @@
+import { insertZuzaluPretixTicket } from "../src/database/queries/zuzalu_pretix_tickets/insertZuzaluPretixTicket";
+import { sqlQuery } from "../src/database/sqlQuery";
+import { ZuconnectTripshaSyncService } from "../src/services/zuconnectTripshaSyncService";
+import { Zupass } from "../src/types";
+import { expectCurrentSemaphoreToBe } from "./semaphore/checkSemaphore";
+import {
+  MOCK_ZUCONNECT_TRIPSHA_URL,
+  badEmptyResponse,
+  badTicketsResponse,
+  getZuconnectMockTripshaServer,
+  goodResponse,
+  makeHandler
+} from "./tripsha/mockTripshaAPI";
+import { testLogin } from "./user/testLoginPCDPass";
+import { overrideEnvironment, testingEnv } from "./util/env";
+import { startTestingApp } from "./util/startTestingApplication";
+
+describe("zuconnect functionality", function () {
+  this.timeout(30_000);
+  let db: Pool;
+  let server: SetupServer;
+  let zuconnectTripshaSyncService: ZuconnectTripshaSyncService;
+  let application: Zupass;
+  let identity: Identity;
+  let user: User;
+  let ticketPCD: EdDSATicketPCD;
+
+  this.afterEach(async () => {
+    server.resetHandlers();
+  });
+
+  this.beforeAll(async () => {
+    await overrideEnvironment(testingEnv);
+    db = await getDB();
+
+    server = getZuconnectMockTripshaServer();
+    server.listen({ onUnhandledRequest: "bypass" });
+
+    application = await startTestingApp({
+      zuconnectTripshaAPI: new ZuconnectTripshaAPI(MOCK_ZUCONNECT_TRIPSHA_URL)
+    });
+
+    if (!application.services.zuconnectTripshaSyncService) {
+      throw new Error("expected there to be a Tripsha sync service");
+    }
+
+    zuconnectTripshaSyncService =
+      application.services.zuconnectTripshaSyncService;
+  });
+
+  this.afterAll(async () => {
+    await stopApplication(application);
+    await db.end();
+    server.close();
+  });
+
+  it("since nobody has logged in yet, all the semaphore groups should be empty", async function () {
+    expectCurrentSemaphoreToBe(application, {
+      p: [],
+      r: [],
+      v: [],
+      o: [],
+      g: []
+    });
+  });
+
+  it("should sync with good API response", async () => {
+    await zuconnectTripshaSyncService.sync();
+    const tickets = await fetchAllZuconnectTickets(db);
+    expect(tickets.length).to.eq(5);
+  });
+
+  it("should soft-delete when tickets do not appear in API response", async () => {
+    server.use(makeHandler({ tickets: goodResponse.tickets.slice(0, 4) }));
+    await zuconnectTripshaSyncService.sync();
+    const tickets = await fetchAllZuconnectTickets(db);
+    expect(tickets.length).to.eq(4);
+
+    const deleted = await sqlQuery(
+      db,
+      `SELECT * FROM zuconnect_tickets WHERE is_deleted = TRUE`
+    );
+    expect(deleted.rowCount).to.eq(1);
+    expect(deleted.rows[0].external_ticket_id).to.eq(
+      goodResponse.tickets[4].id
+    );
+  });
+
+  it("soft-deleted ticket should be un-deleted if ticket appears again in API response", async () => {
+    await zuconnectTripshaSyncService.sync();
```

Comment:
assuming the server goes back to the unmodified handler because of the `afterEach`

Reply:
It does! I'm used to assuming that, but I can make this more explicit if necessary.

Reply:
that's ok, just checking my understanding